### PR TITLE
[WIP] Selective tests

### PR DESCRIPTION
**Pull request by potiuk**
This is a WIP PR for #10507. It has only half of the logic implemented, and I think the final version should be written in python rather than in bash (depending how complex logic we will have). 

----
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
CC: @kaxil  @ashb  You wanted to take a look at it and were keenly interested in it. This one requires #11163  as otherwise there is a risk that tests are not really separated, though this is not a super-hard requirement if we accept the risks involved. 

Due to other obligations I will likely not have time to work on it this week, and since it has some relation with other tasks (#11164 
 and #11163  ) which might or might not be part of 2.0 Alpha release this week, we will have to make some decisions for those. Any help with that is appreciated though.

File: scripts/ci/libraries/_selective_tests.sh [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    echo
+    echo "Changed files"
+    echo
+    echo "${changed_files}" | grep -E "${regexp}" || true
+}
+
+function selective_tests::count_changed_files_matching_patterns() {
+    local regexp
+    regexp=$(selective_tests::build_regexp_string "${@}")
+    count_changed_files=$(echo "${changed_files}" | grep -c -E "${regexp}" || true)
```
Comment (ashb) [inline scripts/ci/libraries/_selective_tests.sh:94]: Missing local for this var

File: scripts/ci/libraries/_selective_tests.sh [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    "^tests/providers"
+)
+readonly AIRFLOW_PROVIDERS_PATTERNS
+
+KUBERNETES_PATTERNS=(
+    "^airflow/kubernetes"
+    "^airflow/providers/cncf"
+    "^tests/kubernetes"
+    "^tests/providers/cncf"
+    "^kubernetes_tests"
```
Comment (ashb) [inline scripts/ci/libraries/_selective_tests.sh:44]: Chart/ folder too?

**Review by ashb — APPROVED** [review]
Looks nice and simple!

Once we merge we should keep cost eye for a week or two when merging changes to makes sure it's running the right tests, but should be ok

**Comment by potiuk** [comment]
This one is very much draft. It jus t shows the direction but the rules there are neither complete nor correct, but I have no time to complete it soon so maybe someone can pick it up if we find this useful

**Comment by potiuk** [comment]
Closing in favour of #11417

**Pull request closed without merging**